packages/deployer/subtasks/validate-storage.js

```diff
@@ -0,0 +1,75 @@
+const { subtask } = require('hardhat/config');
+const logger = require('@synthetixio/core-js/utils/logger');
+
+const { getSlotAddresses, findDuplicateSlots } = require('../internal/ast-helper');
+const { SUBTASK_VALIDATE_STORAGE, SUBTASK_CANCEL_DEPLOYMENT } = require('../task-names');
+
+subtask(SUBTASK_VALIDATE_STORAGE).setAction(async (_, hre) => {
+  logger.subtitle('Validating Storage usage');
+
+  const { contracts } = await _getSourcesAST(hre);
+
+  let errorsFound;
+  errorsFound = _findDuplicateStorageNamespaces(contracts) || errorsFound;
+  errorsFound = _findRegularStorageSlots() || errorsFound;
+  errorsFound = _findInvalidMutationsOnNamespaces() || errorsFound;
+
+  if (errorsFound) {
+    logger.error('Storate usage is not valid');
+    return await hre.run(SUBTASK_CANCEL_DEPLOYMENT);
+  }
+  logger.checked('Namespaces are valid');
+});
+
+async function _getSourcesAST(hre) {
+  const fqns = await hre.artifacts.getAllFullyQualifiedNames();
+
+  const contracts = {};
+  for (const fqn of fqns) {
+    const bi = await hre.artifacts.getBuildInfo(fqn);
+    const split = fqn.split(':');
+    const solPath = split[0];
+    const contractName = split[1];
+    // Hardhat will include the contract name when it builds the json only if any source file has more than one contract.
```

Comment:
nit: can we make this comment shorter (in chars per line) so that it wraps more nicely?

Reply:
done